### ISSUE TITLE
Metal support

### ISSUE THREAD

**Issue by BearishSun** [GameFoundry/bsf#19]
macOS currently uses an OpenGL render backend. This isn't an ideal solution as its OpenGL version is 4.1, meaning compute cannot be used. This requires a variety of rendering effects to be implemented in 4.5 and 4.1 variants, which isn't ideal (although this is something we likely also require for mobile, so maybe not that bad either). Ideally we should add a MetalRenderAPI plugin, and extend XShaderCompiler so it can output Metal shading language code.

**Comment by kattkieru** [comment]
Is there a specific use case for Metal that MoltenVK doesn't cover?

**Comment by BearishSun** [comment]
The main concern is that MolenVK is a wrapper on top of Metal. I haven't used Metal myself yet but from what I've heard there are some considerable differences between it and Vulkan, which means the wrapper might need to do some extra busy-work, making it non-optimal. 

Its probably not the worst thing ever, but since Metal is something users will use if they want maximum performance, having an unnecessary layer on top doesn't feel right. For that reason this is certainly not a huge priority, but will eventually be nice to have.

**Comment by kattkieru** [comment]
I haven't played with it too much yet (still learning Vulkan), but it would at least be a nice stop-gap since you already have a Vulkan backend and you'd get more features on Mac than their OpenGL support offers, plus it could be slipstreamed as a dependency.

Just a thought. :)

**Comment by BearishSun** [comment]
Definitely, I will be adding MoltenVK as soon as I get a chance. Metal support is a more long term goal.

**Comment by crookedbard** [comment]
There is this post about: Apple Rejects iOS App For Using MoltenVK Vulkan, Alleged Non-Public API
[link](https://www.phoronix.com/scan.php?page=news_item&px=Apple-Rejects-iOS-MoltenVK)
Might be interesting to read.

**Comment by christianclavet** [comment]
Read it, MoltenVK was using a non-public interface that is not allowed by apple. There is a pull request that should patch it so it would work again.
[link to article update](https://www.phoronix.com/scan.php?page=news_item&px=MoltenVK-IOSurface-PR)

**Comment by remoe** [comment]
Hi folks. I've tried to build lastest master with MoltenVK. I've changed the file FindVulkan.cmake from:

```set(Vulkan_LIBNAME vulkan)```
to
```set(Vulkan_LIBNAME MoltenVK)```

Then it builds a lot of files, but then it throws this compile error:

```c++
Source/Plugins/bsfVulkanRenderAPI/BsVulkanDevice.cpp:10:
Source/Plugins/bsfVulkanRenderAPI/ThirdParty/vk_mem_alloc.h:1688:14: fatal error: 'malloc.h' file not found
    #include <malloc.h> // for aligned_alloc()
```

I try to find out the reason of this.

btw: this project has a really awesome CMake setup!

**Comment by BearishSun** [comment]
Nice work. You can probably grab most of the render window code from the OpenGL version in `Source\Plugins\bsfGLRenderAPI\MacOS`, and replace any OpenGL calls with Vulkan ones.

**Comment by remoe** [comment]
I will try.

**Comment by remoe** [comment]
I've done a first try. Not all code compiles:

https://github.com/remoe/bsf/commit/e9fae5a04a75946a07b22db4f8c1ba17782ebb96

@BearishSun can you please shortly look at it? 

I don't know much about Vulkan. For example i don't know for what the extension are:

"VK_MVK_MACOS_SURFACE_EXTENSION_NAME"

https://github.com/remoe/bsf/commit/e9fae5a04a75946a07b22db4f8c1ba17782ebb96#diff-4fe8409ac8a139179166584cc5ff8657R152

is it similar to "VK_KHR_WIN32_SURFACE_EXTENSION_NAME" on windows?

**Comment by BearishSun** [comment]
Yes that looks pretty close to what's its supposed to be. 

Indeed `VK_MVK_MACOS_SURFACE_EXTENSION_NAME` is just a macOS version of that Win32 extension. Here's the window-system-specific part of the Vulkan spec that explains it: https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#wsi
